neo4j-jdbc-translator/sparkcleaner/src/main/java/org/neo4j/jdbc/translator/sparkcleaner/SparkSubqueryCleaningTranslator.java

```
@@ -0,0 +1,91 @@
+
+import org.antlr.v4.runtime.CharStreams;
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.neo4j.cypher.internal.parser.v5.Cypher5Lexer;
+import org.neo4j.cypher.internal.parser.v5.Cypher5Parser;
+import org.neo4j.jdbc.translator.spi.Translator;
+
+/**
+ * This translator will search for the occurrence of {@literal SPARK_GEN_SUBQ_} in a query
+ * before passing it on towards to the next translator. If such a literal is found, a
+ * check if it might be a spark subquery containing cypher instead of SQL is performed.
+ * The check is performed by parsing it with jOOQ as well.
+ *
+ * @author Michael J. Simons
+ * @since 6.2.0
+ */
+final class SparkSubqueryCleaningTranslator implements Translator {
+
+	private static final Pattern SUBQUERY_PATTERN = Pattern
+		.compile("(?ims)SELECT\\s+\\*\\s+FROM\\s+\\((.*?)\\)\\s+SPARK_GEN_SUBQ_0.*");
+
+	private final int precedence;
+
+	SparkSubqueryCleaningTranslator(int precedence) {
+		this.precedence = precedence;
+	}
+
+	@Override
+	public int getOrder() {
+		return this.precedence;
+	}
+
+	@Override
+	public String translate(String statement, DatabaseMetaData optionalDatabaseMetaData) {
+		if (!mightBeASparkQuery(statement)) {
+			return null;
+		}
+
+		var extractedSubquery = extractSubquery(statement);
+		return extractedSubquery.filter(this::canParseAsCypher).map(v -> """
+				/*+ NEO4J FORCE_CYPHER */
+				CALL {%s} RETURN * LIMIT 1
+				""".formatted(v).strip()).orElse(null);
+	}
+
+	static boolean mightBeASparkQuery(String statement) {
+		return statement != null && statement.toUpperCase(Locale.ROOT).contains("SPARK_GEN_SUBQ");
+	}
+
+	static Optional<String> extractSubquery(String statement) {
+		var matcher = SUBQUERY_PATTERN.matcher(statement);
+		if (!matcher.matches()) {
+			return Optional.empty();
+		}
+		return Optional.of(matcher.group(1).trim());
+	}
+
+	boolean canParseAsCypher(String statement) {
+
+		var tokens = new CommonTokenStream(new Cypher5Lexer(CharStreams.fromString(statement)));
```

FYI if you want full Cypher support you need to replace unicode escape codes before the lexer, that's not part of the antlr grammar but is a part of Cypher. Understand if that's not important to your use case though. See https://github.com/neo-technology/neo4j/blob/dev/public/community/cypher/front-end/parser/common/antlr-ast-common/src/main/java/org/neo4j/cypher/internal/parser/lexer/UnicodeEscapeReplacementReader.java.

---

You can also use `parser.setErrorHandler(new BailErrorStrategy())` to fail fast. By default antlr tries to recover from errors by rewriting the query.

---

Thanks a lot!